src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,278 @@
-        let stack = entry.stack;
-        let location = getLocation(node);
-        stack.unshift(location.id as number);
-        if (node.hitCount > 0) {
-          const sample = new perftools.profiles.Sample({
-            locationId: stack,
-            value: [node.hitCount, node.hitCount * intervalMicros]
-          });
-          samples.push(sample);
-        }
-        for (let child of node.children) {
-          entries.push({node: child, stack: stack.slice(0)});
-        }
+  createAllocationValueType(): perftools.profiles.ValueType {
+    return new perftools.profiles.ValueType(
+        {type: this.getIndexOrAdd('space'), unit: this.getIndexOrAdd('bytes')});
+  }
+}
+
+/**
+ * Takes v8 profile and populates sample, location, and function fields of
+ * profile.proto.
+ *
+ * @param profile - profile.proto with empty sample, location, and function
+ * fields.
+ * @param root - root of v8 profile tree describing samples to be appended
+ * to profile.
+ * @param appendToSample - function which converts entry to sample(s)  and
+ * appends these to end of an array of samples.
+ * @param stringTable - string table for the existing profile.
+ */
+function serialize(
+    profile: perftools.profiles.IProfile, root: ProfileNode,
+    appendToSample: AppendEntryToSamples, stringTable: StringTable) {
+  const samples: Array<perftools.profiles.Sample> = [];
+  const locations: Array<perftools.profiles.Location> = [];
+  const functions: Array<perftools.profiles.Function> = [];
+  const locationMap: Map<number, perftools.profiles.Location> = new Map();
+  const functionMap: Map<number, perftools.profiles.Function> = new Map();
+  const functionIdMap = new Map<string, number>();
+  const locationIdMap = new Map<string, number>();
+
+  const entries: Entry[] = root.children.map((n) => ({node: n, stack: []}));
+  while (entries.length > 0) {
+    const entry = entries.pop();
+    if (entry !== undefined) {
+      const node = entry.node;
+      const stack = entry.stack;
+      const location = getLocation(node);
+      stack.unshift(location.id as number);
+      appendToSample(entry, samples);
+      for (let child of node.children) {
+        entries.push({node: child, stack: stack.slice(0)});
       }
     }
   }
 
-  function getLocation(node: TimeProfileNode): perftools.profiles.Location {
-    const id = node.callUid;
-    let location = locationMap.get(id);
-    if (location !== undefined) {
-      return location;
+  profile.sample = samples;
+  profile.location = locations;
+  profile.function = functions;
+  profile.stringTable = stringTable.strings;
+
+  function getLocation(node: ProfileNode): perftools.profiles.Location {
+    const key = {
+      functionKey: {name: node.name, scriptId: node.scriptId},
+      lineNumber: node.lineNumber,
+      columnNumber: node.columnNumber
+    };
+    const keyStr = JSON.stringify(key);
+    let id = locationIdMap.get(keyStr);
+    if (id !== undefined && id <= locations.length) {
```

Comment:
Why do you have the " && id <= locations.length" check? That's an invariant, no?